### PR TITLE
topsql: make topsql enable only be controlled by pub/sub sink

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### What problem does this PR solve?

Issue Number: close #xxx

make topsql enable only be controlled by pub/sub sink

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

**Review by zhongzc — APPROVED** [review]

File: sessionctx/variable/sysvar.go [hunk 1]
```diff
@@ -1246,12 +1246,9 @@ var defaultSysVars = []*SysVar{
...
 	// variable for top SQL feature.
-	{Scope: ScopeGlobal, Name: TiDBEnableTopSQL, Value: BoolToOnOff(topsqlstate.DefTiDBTopSQLEnable), Type: TypeBool, Hidden: true, AllowEmpty: true, GetGlobal: func(s *SessionVars) (string, error) {
-		return BoolToOnOff(topsqlstate.GlobalState.Enable.Load()), nil
-	}, SetGlobal: func(vars *SessionVars, s string) error {
-		topsqlstate.GlobalState.Enable.Store(TiDBOptOn(s))
-		return nil
-	}, GlobalConfigName: GlobalConfigEnableTopSQL},
+	// TopSQL enable only be controlled by TopSQL pub/sub sinker.
+	// This global variable only uses to update the global config which store in PD(ETCD).
+	{Scope: ScopeGlobal, Name: TiDBEnableTopSQL, Value: BoolToOnOff(topsqlstate.DefTiDBTopSQLEnable), Type: TypeBool, Hidden: true, AllowEmpty: true, GlobalConfigName: GlobalConfigEnableTopSQL},
```
Comment (breeswish) [inline sessionctx/variable/sysvar.go:1251]: I think it should not be hidden any more.

File: util/topsql/state/state.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ const (
 
 // GlobalState is the global Top-SQL state.
 var GlobalState = State{
-	Enable:                atomic.NewBool(DefTiDBTopSQLEnable),
+	enable:                atomic.NewBool(DefTiDBTopSQLEnable),
```
Comment (breeswish) [inline util/topsql/state/state.go:30]: Why does the state have a default? I think it should be always false initially, in all cases.

**Review by breeswish — APPROVED** [review]
The rest LGTM

**Comment by breeswish** [comment]
You have made a mistake that the variable and state shares the same default value. However they are totally different things so that they shoud have different default constants.

**Comment by crazycs520** [comment]
> You have made a mistake that the variable and state share the same default value. However they are totally different things so that they should have different constants for controlling the default value.

Nice catch! resolved.

**Review by breeswish — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]

**Comment by breeswish** [comment]
@crazycs520 The unit test is failed. Please fix it.

File: executor/set_test.go [hunk 3]
```diff
@@ -1602,5 +1602,5 @@ func (s *testSerialSuite) TestSetTopSQLVariables(c *C) {
 
 	// Test for hide top sql variable in show variable.
```
Comment (breeswish) [inline executor/set_test.go:1603]: Please change the comment.

**Pull request merged**